File: 2023/03. Gear Ratios/c++/one/src/main.cpp
#include <array>
#include <cctype>
#include <cstddef>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <iterator>
#include <numeric>
#include <string>
#include <unordered_map>
#include <vector>

int main(const int argc, const char** const argv)
{
    if (argc != 2)
    {
        std::cerr << "Usage: " << argv[0] << " <INPUT_FILE>\n";
        return 1;
    }

    std::ifstream infile{argv[1]};
    if (!infile.is_open())
    {
        std::cerr << "ERROR! Could not open input file " << std::quoted(argv[1]) << '\n';
        return 1;
    }

    // Map of (pointer to start of number, numeric value)
    // Used to prevent the same instance of a number from being added multiple times while still allow the same numeric value to be duplicated
    std::unordered_map<char*, int> partNumbers;

    // Store the grid loaded from the file
    std::vector<std::vector<char>> grid;

    // Lambda to check if a given tile is part of a number and then add that number to partNumbers
    const auto processNeighbor{[&partNumbers, &grid](const std::size_t x, const std::size_t y)
    {
        // No needed to check less than zero because std::size_t is unsigned
        if (y >= grid.size() || x >= grid[y].size())
            return;

        if (std::isdigit(grid[y][x]))
        {
            // Find start of number
            auto first{grid[y].begin() + static_cast<std::ptrdiff_t>(x)};
            while (first != grid[y].begin() && std::isdigit(*std::prev(first)))
            {
                --first;
            }

            // Find end of number
            auto last{grid[y].begin() + static_cast<std::ptrdiff_t>(x)};
            while (last != std::prev(grid[y].end()) && std::isdigit(*std::next(last)))
            {
                ++last;
            }

            // Parse number and add to map
            const int partNumber{std::stoi(std::string{first, last + 1})};
            partNumbers.insert({first.base(), partNumber});
        }
    }};

    // Read each line
    for (std::string line; std::getline(infile, line);)
    {
        grid.emplace_back(line.begin(), line.end());
    }

    // Iterate over all tiles looking for symbols
    for (std::size_t y{}; y < grid.size(); ++y)
    {
        for (std::size_t x{}; x < grid[y].size(); ++x)
        {
            const char c{grid[y][x]};

            if (c != '.' && std::ispunct(c))
            {
                processNeighbor(x, y - 1); // North
                processNeighbor(x, y + 1); // South
                processNeighbor(x + 1, y); // East
                processNeighbor(x - 1, y); // West
                processNeighbor(x + 1, y - 1); // North-East
                processNeighbor(x + 1, y + 1); // South-East
                processNeighbor(x - 1, y + 1); // South-West
                processNeighbor(x - 1, y - 1); // North-West
            }
        }
    }

    // Sum part numbers
    int partNumberSum{};
    for (const auto&[_, partNumber] : partNumbers)
    {
        partNumberSum += partNumber;
    }

    std::cout << "Result: " << partNumberSum << '\n';

    return 0;
}

File: 2023/03. Gear Ratios/rust/one/src/main.rs
use std::collections::HashMap;
use std::env;
use std::fs;

// Add the part number containg x, y in grid to part_numbers of present
fn process_neighbor(
    grid: &Vec<&str>,
    part_numbers: &mut HashMap<(usize, usize), i32>,
    x: usize,
    y: usize,
) {
    // No need to check below zero because usize is unsigned
    if y >= grid.len() || x >= grid[y].len() {
        return;
    }

    let row = grid[y];
    let c = row.chars().nth(x).expect("Failed to index gird");

    if c.is_ascii_digit() {
        // Find first digit of part number
        let begin = match row[..x].rfind(|c: char| !c.is_ascii_digit()) {
            Some(index) => index + 1,
            None => 0,
        };

        // Find past-the-last digit of part number
        let end = match row[x..].find(|c: char| !c.is_ascii_digit()) {
            Some(index) => x + index,
            None => row.len(),
        };

        // Convert part number to i32 and insert into map
        let part_number: i32 = row[begin..end].parse().expect(&format!(
            "Failed to parse part number \"{}\"",
            &row[begin..end]
        ));

        // Key is coordinate of the start of part number string to allow duplicates of same string in different location
        part_numbers.insert((begin, y), part_number);
    }
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        eprintln!("Usage: {} <INPUT_FILE>", &args[0]);
    }

    let mut part_numbers = HashMap::new();

    let contents = fs::read_to_string(&args[1]).expect("Failed to read input file");
    let grid: Vec<&str> = contents.lines().collect();

    for (y, row) in grid.iter().enumerate() {
        for (x, _c) in row
            .char_indices()
            .filter(|(_x, c)| c.is_ascii_punctuation() && *c != '.') // Only iterate over symbols
        {
            process_neighbor(&grid, &mut part_numbers, x, y.wrapping_sub(1)); // North
            process_neighbor(&grid, &mut part_numbers, x, y.wrapping_add(1)); // South
            process_neighbor(&grid, &mut part_numbers, x.wrapping_add(1), y); // East
            process_neighbor(&grid, &mut part_numbers, x.wrapping_sub(1), y); // West

            // North-East
            process_neighbor(
                &grid,
                &mut part_numbers,
                x.wrapping_add(1),
                y.wrapping_sub(1),
            );
            // South-East
            process_neighbor(
                &grid,
                &mut part_numbers,
                x.wrapping_add(1),
                y.wrapping_add(1),
            );
            // South-West
            process_neighbor(
                &grid,
                &mut part_numbers,
                x.wrapping_sub(1),
                y.wrapping_add(1),
            );
            // North-West
            process_neighbor(
                &grid,
                &mut part_numbers,
                x.wrapping_sub(1),
                y.wrapping_sub(1),
            );
        }
    }

    let part_number_sum: i32 = part_numbers.values().sum();
    println!("Result: {part_number_sum}");
}
